### PR TITLE
Deprecate `block_diag` from math module in favor of PyTensor

### LINKED ISSUE #7085: Remove `block_diag` from `pymc.math` in favor of alias to `pytensor.tensor.slinalg.block_diag`
### Description

`block_diag` is a core math function that should live in `pytensor` and be aliased in `pm.math`, rather than being directly defined there. https://github.com/pymc-devs/pytensor/pull/576 moves the block_diag code, along with relevant support code (`pm.math.ix`, `pm.math.largest_common_dtype`, into pytensor. The PyMC function should be replaced with an alias to that new function once merged.

In addition, the pytensor function is going to more closely follow the scipy API, which means that the function signature will change from `pm.math.block_diag(matrices)` to `pm.math.block_diag(*matrices)`. Scipy also has separate functions for the sparse and dense cases, which I think we should have as well (rather than a single function). This necessitates a depreciation warning and a temporary helper function to dispatch to either `pytensor.tensor.slinalg.block_diag` or `pytensor.sparse.block_diag`

### PR DESCRIPTION
**Pull request by AryanNanda17**
The new code uses the sparse parameter to determine whether to call the sparse block_diag function (pytensor.sparse.basic.block_diag) or the dense version (pt.slinalg.block_diag).
## Related Issue
- [x] Closes #7085

## Checklist
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature/enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc--7132.org.readthedocs.build/en/7132/

### CODE REVIEW

**Comment by AryanNanda17** [comment]
@ricardoV94 and @jessegrabowski please look into the PR and suggest if any changes are required.

**Comment by AryanNanda17** [comment]
What am I doing wrong here, the Codecov/patch test failed?

**Comment by ricardoV94** [comment]
@AryanNanda17 I asked the person commenting earlier on the original issue if they would be interested in working on this issue as a courtesy. If you see an issue you would like to work on where someone else seemed to be interested recently please confirm with them first if they are happy to let you take over.

Our official policy is opening a PR first is what counts, but for these small non critical issues we would like to give room for everyone to take their chance. There are many harder issues that we need help with for which there's no risk of multiple people being interested at the same time.

**Comment by ricardoV94** [comment]
@AryanNanda17 

These beginner friendly ones don't seem to have anyone interested and may be good?
- https://github.com/pymc-devs/pymc/issues/6985
- https://github.com/pymc-devs/pymc/issues/6686 (there's a stale PR, fine to take over)

We also have many good issue in the related repos: PyTensor and PyMC-Experimental that tend to fall more out of the radar:
- https://github.com/pymc-devs/pymc-experimental
- https://github.com/pymc-devs/pytensor/issues?q=is%3Aopen+is%3Aissue+label%3A%22beginner+friendly%22

File: pymc/math.py [hunk 1]
```diff
@@ -577,4 +528,8 @@ def block_diagonal(matrices, sparse=False, format="csr"):
     """
     if len(matrices) == 1:  # graph optimization
         return matrices[0]
-    return BlockDiagonalMatrix(sparse=sparse, format=format)(*matrices)
+
```
Comment (ricardoV94) [inline pymc/math.py]: Can you add a FutureWarning mentioning this is deprecated in favor of the pytensor function, so we can remove this thin wrapper sometime in the future?

**Comment by AryanNanda17** [comment]
@ricardoV94, I made the changes please checkout.

File: pymc/math.py [hunk 2]
```diff
@@ -576,5 +527,13 @@ def block_diagonal(matrices, sparse=False, format="csr"):
     matrix
     """
     if len(matrices) == 1:  # graph optimization
+        warnings.warn(
```
Comment (ricardoV94) [inline pymc/math.py]: Any use of the function should issue a warning, not just this

Comment (AryanNanda17) [reply pymc/math.py]: Alright!

File: pymc/math.py [hunk 3]
```diff
@@ -575,6 +526,14 @@ def block_diagonal(matrices, sparse=False, format="csr"):
     -------
     matrix
     """
+    warnings.warn(
+        "The behavior of block_diagonal when only one matrix is provided is deprecated. Use pytensor function instead.",
+        FutureWarning,
+        stacklevel=2,
+    )
```
Comment (ricardoV94) [inline pymc/math.py]:
```suggestion
    warnings.warn(
        "`pymc.math.block_diagonal` is deprecated in favor of `pytensor.tensor.linalg.block_diag` and `pytensor.sparse.block_diag`. This function will be removed in a future release.",
    )
```

**Comment by AryanNanda17** [comment]
Thanks for looking into the issue @ricardoV94.

File: pymc/math.py [hunk 4]
```diff
@@ -513,55 +510,9 @@ def batched_diag(C):
...
-
-    def infer_shape(self, fgraph, nodes, shapes):
-        first, second = zip(*shapes)
-        return [(pt.add(*first), pt.add(*second))]
-
-
-def block_diagonal(matrices, sparse=False, format="csr"):
-    r"""See scipy.sparse.block_diag or
-    scipy.linalg.block_diag for reference
+def block_diagonal(*matrices, sparse=False, format="csr"):
```
Comment (ricardoV94) [inline pymc/math.py]: We should keep the old signature, missed this before:

```suggestion
def block_diagonal(matrices, sparse=False, format="csr"):
```

Comment (AryanNanda17) [reply pymc/math.py]: Updated it!

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Thanks @AryanNanda17

**Pull request merged**